storage/innobase/include/clone0desc.h

```diff
@@ -0,0 +1,650 @@
+  CLONE_SNAPSHOT_REDO_COPY -right> CLONE_SNAPSHOT_DONE
+  CLONE_SNAPSHOT_DONE -down-> [*] : Destroy snapshot
+@enduml
+
+Clone Type: HA_CLONE_PAGE: Not implemented
+*/
+enum Snapshot_State : uint32_t {
+  /** Invalid state */
+  CLONE_SNAPSHOT_NONE = 0,
+
+  /** Initialize state when snapshot object is created */
+  CLONE_SNAPSHOT_INIT,
+
+  /** Snapshot state while transferring files. */
+  CLONE_SNAPSHOT_FILE_COPY,
+
+  /** Snapshot state while transferring pages. */
+  CLONE_SNAPSHOT_PAGE_COPY,
+
+  /** Snapshot state while transferring redo. */
+  CLONE_SNAPSHOT_REDO_COPY,
+
+  /** Snapshot state at end after finishing transfer. */
+  CLONE_SNAPSHOT_DONE
+};
+
+/** Total number of data transfer stages in clone. */
+const size_t CLONE_MAX_TRANSFER_STAGES = 3;
+
+/** Choose lowest descriptor version between reference locator
+and currently supported version.
+@param[in]      ref_loc reference locator
+@return chosen version */
+uint choose_desc_version(const byte *ref_loc);
+
+/** Check if clone locator is valid
+@param[in]      desc_loc        serialized descriptor
+@param[in]      desc_len        descriptor length
+@return true, if valid locator */
+bool clone_validate_locator(const byte *desc_loc, uint desc_len);
+
+/** Clone descriptors contain meta information needed for applying cloned data.
+These are PODs with interface to serialize and deserialize them. */
+enum Clone_Desc_Type {
+  /** Logical pointer to identify a clone operation */
+  CLONE_DESC_LOCATOR = 1,
+
+  /** Metadata for a Task/Thread for clone operation */
+  CLONE_DESC_TASK_METADATA,
+
+  /** Information for snapshot state */
+  CLONE_DESC_STATE,
+
+  /** Metadata for a database file */
+  CLONE_DESC_FILE_METADATA,
+
+  /** Information for a data block */
+  CLONE_DESC_DATA,
+
+  /** Must be the last member */
+  CLONE_DESC_MAX
+};
+
+/** Header common to all descriptors. */
+struct Clone_Desc_Header {
+  /** Descriptor version */
+  uint m_version;
+
+  /** Serialized length of descriptor in bytes */
+  uint m_length;
+
+  /** Descriptor type */
+  Clone_Desc_Type m_type;
+
+  /** Serialize the descriptor header: Caller must allocate
+  the serialized buffer.
+  @param[out]   desc_hdr        serialized header */
+  void serialize(byte *desc_hdr);
+
+  /** Deserialize the descriptor header.
+  @param[in]    desc_hdr        serialized header
+  @param[in]    desc_len        descriptor length
+  @return true, if successful. */
+  bool deserialize(const byte *desc_hdr, uint desc_len);
+};
+
+/** Task information in clone operation. */
+struct Clone_Task_Meta {
+  /** Index in task array. */
+  uint m_task_index;
+
+  /** Current chunk number reserved by the task. */
+  uint m_chunk_num;
+
+  /** Current block number that is already transferred. */
+  uint m_block_num;
+};
+
+/** Map for current block number for unfinished chunks. Used during
+restart from incomplete clone operation. */
+using Chunk_Map = std::map<uint32_t, uint32_t>;
```

Comment:
What is an incomplete clone operation and how does the restart happen? Could the comment mention some relevant function names or SQL statements? What is protecting the `Chunk_Info::m_incomplete_chunks`? Why do we have this alias at all?